packages/build-info/src/mutex.ts

```diff
@@ -0,0 +1,88 @@
+/**
+ * A lock for synchronizing async operations.
+ * Use this to protect a critical section
+ * from getting modified by multiple async operations
+ * at the same time.
+ */
+export class Mutex {
+  /**
+   * When multiple operations attempt to acquire the lock,
+   * this queue remembers the order of operations.
+   */
+  private _queue: {
+    resolve: (release: ReleaseFunction) => void
+  }[] = []
+
+  private _isLocked = false
+
+  /**
+   * Wait until the lock is acquired.
+   * @returns A function that releases the acquired lock.
+   */
+  acquire() {
+    return new Promise<ReleaseFunction>((resolve) => {
+      this._queue.push({ resolve })
+      this._dispatch()
+    })
+  }
+
+  /**
+   * Enqueue a function to be run serially.
+   *
+   * This ensures no other functions will start running
+   * until `callback` finishes running.
+   * @param callback Function to be run exclusively.
+   * @returns The return value of `callback`.
+   */
+  async runExclusive<T>(callback: () => Promise<T>) {
```

Comment:
could the same be achieved with just `p-limit` using `1` as concurrency? It seems like `p-limit` is already indirect dependency of `build-info`:
```
npm list p-limit
netlify-build@0.0.0 /Users/misiek/dev/netlify-build
└─┬ @netlify/build-info@7.14.4 -> ./packages/build-info
  ├─┬ @bugsnag/source-maps@2.3.3
  │ └─┬ read-pkg-up@7.0.1
  │   └─┬ find-up@4.1.0
  │     └─┬ locate-path@5.0.0
  │       └─┬ p-locate@4.1.0
  │         └── p-limit@2.3.0
  ├─┬ @vitest/coverage-c8@0.33.0
  │ └─┬ c8@7.14.0
  │   └─┬ find-up@5.0.0
  │     └─┬ locate-path@6.0.0
  │       └─┬ p-locate@5.0.0
  │         └── p-limit@3.1.0
  ├─┬ find-up@6.3.0
  │ └─┬ locate-path@7.2.0
  │   └─┬ p-locate@6.0.0
  │     └── p-limit@4.0.0 deduped
  └─┬ vitest@0.34.6
    └─┬ @vitest/runner@0.34.6
      └── p-limit@4.0.0
```

(no strong preference here, just checking)